Allow shared parent mock to be passed to Device.connect

jsouter opened the pull request:
Addresses https://github.com/bluesky/ophyd-async/issues/336

Allows a Device to accept a unittest.Mock or a boolean.
If True passed, creates a new Mock()
if Mock() passed, finds any child mocks set with attach_mock if they exist, else passes down mock=True to children.

Will push some tests for this

edit: now realising that I misunderstood the requirements, will fix quickly!

DominicOram reviewed (approved):
I think this looks about right. As you say a bunch of tests confirming all the use cases would be great

jsouter commented:
Would have to call attach_mock after each child device is connected as that is when the signal backend gets set to MockSignalBackend and the _backend.put_mock can be accessed. Tom has said he may touch how that is handled soon so I will wait for that to be done first.

coretl commented:
#594 is sufficiently advanced that you can work on top of it to see if you can apply this on top of it

coretl commented on src/ophyd_async/core/_device.py:
```diff
@@ -89,29 +90,38 @@ async def connect(
+            for child_name, child_device in self.children():
+                # if mock is Mock instance, get child Mock if exists
+                # else pass down boolean
+                child_mock = getattr(mock, child_name, True) if mock else mock
```
mock should always be a `Mock` here by now
```suggestion
                child_mock = getattr(mock, child_name) if mock else mock
```

jsouter commented:
I've rebased on top of https://github.com/bluesky/ophyd-async/pull/594, still a WIP but it appears to work with the existing tests. I will work on porting over the new tests I added in the force-pushed-over commits here

jsouter commented on src/ophyd_async/core/_mock_signal_utils.py:
```diff
@@ -1,8 +1,9 @@
-from unittest.mock import AsyncMock
+from unittest.mock import AsyncMock, Mock
 
+from ._device import Device
```
I think this is mostly in a sensible state now, depending on if we're happy with get_mock in _mock_signal_utils accepting Device | Signal

jsouter commented:
There's a few places where the signature for connect is still `mock: bool`: in the Connectable Protocol, PviDeviceConnector and TangoDeviceConnector etc, I assume we want to make it uniform?

coretl commented on src/ophyd_async/core/_device.py:
```diff
@@ -106,12 +108,15 @@ async def connect(
         )
+        if mock is True:
+            mock = Mock()  # create a new Mock if one not provided
+        self.mock = mock
```
Hmm, having seen this I'm inclined to go back to having a global to store all the mocks and MockSignalBackends in, like we had ages ago:
https://github.com/bluesky/ophyd-async/blob/54deda30cae717472d1020bd5a67489380960ad8/src/ophyd_async/core/signal.py#L24

jsouter replied:
Okay, do we want the mapping to be

- dict[Signal, MockSignalBackend]
- dict[MockSignalBackend, Mock]
- dict[Signal, Mock]

coretl replied:
```python
_mock_signal_backends: dict[Signal, MockSignalBackend]  # used for `set_mock_value`
_device_mocks: dict[Device, Mock]  # used for `get_mock`
```

coretl commented on src/ophyd_async/core/_mock_signal_utils.py:
```diff
@@ -47,6 +48,12 @@ def get_mock_put(signal: Signal) -> AsyncMock:
 
 
+def get_mock(device: Device | Signal) -> Mock:
+    if isinstance(device, Signal):
+        return _get_mock_signal_backend(device).mock
+    return device.mock
```
Then this would be:
```python
    return _device_mocks[device]
```

coretl commented on src/ophyd_async/core/_mock_signal_utils.py:
```diff
@@ -47,6 +48,12 @@ def get_mock_put(signal: Signal) -> AsyncMock:
+
+
 def reset_mock_put_calls(signal: Signal):
     backend = _get_mock_signal_backend(signal)
```
And this would be:
```python
    return _mock_signal_backends[signal]
```

coretl commented on src/ophyd_async/core/_mock_signal_backend.py:
```diff
@@ -13,7 +13,11 @@
+    def __init__(
+        self,
+        initial_backend: SignalBackend[SignalDatatypeT],
+        mock: bool | Mock = True,
```
I think we should create the Mock before it gets passed in:
```suggestion
        mock: Mock,
```

coretl commented on src/ophyd_async/core/_device.py:
```diff
@@ -121,6 +129,10 @@ async def connect(
         # Wait for it to complete
         await self._connect_task
 
+    def __hash__(self):
+        # to allow Devices to be used as dict keys
+        return hash(id(self))
+
```
I think Devices are already hashable...

jsouter replied:
You're right! Will remove that now

coretl commented on src/ophyd_async/core/_device.py:
```diff
@@ -175,6 +187,9 @@ def children(self) -> Iterator[tuple[str, Device]]:
         for key, child in self._children.items():
             yield str(key), child
 
+    def __hash__(self):  # to allow DeviceVector to be used as dict keys
+        return hash(id(self))
+
```
But you're right that we need this to make DeviceVector hashable

DominicOram commented on tests/epics/demo/test_demo.py:
```diff
@@ -199,6 +199,28 @@ async def test_retrieve_mock_and_assert(mock_mover: demo.Mover):
     )
 
 
+async def test_mock_with_Mock():
```
Nit: This name is a bit too implementation focussed IMO, something like `test_mocks_in_device_share_parent` would be more descriptive of the usecase

coretl commented on src/ophyd_async/core/_mock_signal_backend.py:
```diff
@@ -27,6 +31,11 @@ def __init__(self, initial_backend: SignalBackend[SignalDatatypeT]) -> None:
             )
+
+        # use existing Mock if provided
+        self.mock = mock
+        self.mock.attach_mock(AsyncMock(name="put", spec=Callable), "put")
```
Could:
```suggestion
        self.mock = mock
        self.put_mock = AsyncMock(name="put", spec=Callable)
        self.mock.attach_mock(put_mock, "put")
```

coretl commented on src/ophyd_async/core/_mock_signal_backend.py:
```diff
@@ -40,7 +49,7 @@ async def connect(self, timeout: float) -> None:
 
     @cached_property
     def put_mock(self) -> AsyncMock:
-        return AsyncMock(name="put", spec=Callable)
+        return self.mock.put
```
Then you could delete this cached property...

coretl commented on src/ophyd_async/core/_mock_signal_utils.py:
```diff
@@ -1,20 +1,19 @@
-    return backend
+    assert (
+        signal in _mock_signal_backends
+    ), "No `MockSignalBackend` registered for signal."
```
```suggestion
    ), f"Signal {signal} not connected in mock mode"
```

coretl commented on tests/epics/demo/test_demo.py:
```diff
@@ -199,6 +199,28 @@ async def test_retrieve_mock_and_assert(mock_mover: demo.Mover):
+    await mock_mover.setpoint.set(10)
+    get_mock_put(mock_mover.setpoint).assert_called_once_with(10, wait=ANY)
+
+    await mock_mover.velocity.set(100)
+    await mock_mover.setpoint.set(67)
+
+    mock.assert_has_calls(
+        [
+            call.velocity.put(100, wait=True),
+            call.setpoint.put(67, wait=True),
+        ]
+    )
```
would prefer to be explicit about the previous calls (assert_has_calls does a subset check):
```suggestion
    mock.reset_mock()
    await mock_mover.velocity.set(100)
    await mock_mover.setpoint.set(67)

    assert mock.call_args_list == [
        call.velocity.put(100, wait=True),
        call.setpoint.put(67, wait=True),
    ]
```

coretl commented:
Shall I merge?
